File: solutions/2026-07-13/python/2026-07-13-sequential-digits.py
```python
# LeetCode Problem: Sequential Digits
# Link: https://leetcode.com/problems/sequential-digits/
#
# Problem Summary: Find all numbers within a given range [low, high] that have digits increasing sequentially by 1.
#
# Approach:
# The core idea is to generate all possible sequential digit numbers and then filter them based on the given range [low, high].
# Since the maximum value for 'high' is 10^9, the maximum length of a sequential digit number we need to consider is 9 (e.g., 123456789).
# We can pre-generate all possible sequential digit numbers.
#
# There are 9 possible starting digits (1 through 9).
# For each starting digit, we can form sequential numbers of increasing length.
#
# Example:
# Starting with '1':
#   Length 2: 12
#   Length 3: 123
#   ...
#   Length 9: 123456789
#
# Starting with '2':
#   Length 2: 23
#   Length 3: 234
#   ...
#   Length 8: 23456789
#
# And so on, until starting with '9'.
#
# We can iterate through possible lengths (from 2 to 9) and starting digits (from 1 to 9-length+1).
# Alternatively, a simpler way is to iterate through the 10 available digits (0-9) and use them as a string to build sequential numbers.
#
# For example, using the string "123456789":
#   We can take all substrings of length 2 to 9.
#   Substrings of length 2: "12", "23", "34", ..., "89"
#   Substrings of length 3: "123", "234", ..., "789"
#   ...
#   Substrings of length 9: "123456789"
#
# After generating all these numbers, we convert them to integers and filter them to keep only those within the [low, high] range.
# Finally, we sort the resulting list.
#
# Time Complexity:
# The number of sequential digit numbers is relatively small.
# The maximum length of a sequential digit number is 9.
# For a starting digit `d` and length `l`, the last digit is `d + l - 1`. The maximum last digit is 9. So, `d + l - 1 <= 9`.
# The total number of sequential numbers is bounded.
# For length 2: 8 numbers (12, 23, ..., 89)
# For length 3: 7 numbers (123, 234, ..., 789)
# ...
# For length 9: 1 number (123456789)
# The total is 8 + 7 + 6 + 5 + 4 + 3 + 2 + 1 = 36.
# Generating these numbers by iterating through substrings of "123456789" takes O(L^2) time, where L is the length of the string (L=9). So, O(9^2) which is constant.
# Filtering and sorting these 36 numbers takes O(N log N) where N is the number of sequential numbers (N=36). This is also constant.
# Therefore, the overall time complexity is O(1) because the number of operations is bounded by a constant, independent of the input range 'low' and 'high' beyond their maximum constraints.
#
# Space Complexity:
# We store the generated sequential digit numbers in a list. The maximum number of such integers is constant (36).
# The output list will also contain at most 36 elements.
# Therefore, the space complexity is O(1).

class Solution:
    def sequentialDigits(self, low: int, high: int) -> list[int]:
        # String containing all digits 0-9 for easy substring generation.
        # We only need '1' through '9' to generate sequential numbers.
        digits = "123456789"
        
        # List to store all valid sequential digit numbers found.
        sequential_numbers = []
        
        # Iterate through all possible lengths of sequential numbers.
        # Minimum length is 2 (e.g., 12), maximum length is 9 (123456789).
        for length in range(2, 10):
            # Iterate through all possible starting positions in the 'digits' string.
            # The starting position 'i' determines the first digit of the sequential number.
            # The last possible starting position for a given length is when the sequential number
            # ends exactly at '9'. For example, if length is 3, the last starting digit can be '7' (789).
            # So, i can go up to len(digits) - length.
            for i in range(len(digits) - length + 1):
                # Extract the substring of the desired length starting from index 'i'.
                # This substring represents a sequential digit number.
                sub = digits[i : i + length]
                
                # Convert the substring to an integer.
                num = int(sub)
                
                # Check if the generated sequential number falls within the given range [low, high].
                if low <= num <= high:
                    # If it's within the range, add it to our list.
                    sequential_numbers.append(num)
                    
        # The problem requires the output list to be sorted.
        # Since we generated numbers in a somewhat ordered fashion (by length and then by start digit),
        # sorting ensures the final output meets the requirement.
        sequential_numbers.sort()
        
        # Return the sorted list of sequential digit numbers within the range.
        return sequential_numbers

```

File: solutions/2026-07-13/javascript/2026-07-13-sequential-digits.js
/**
 * @fileoverview LeetCode Problem: Sequential Digits
 * @problem_summary Returns a sorted list of all integers in a given range [low, high] that have sequential digits.
 * @problem_link https://leetcode.com/problems/sequential-digits/
 *
 * @approach
 * The problem asks for numbers with sequential digits within a given range.
 * Sequential digits mean each digit is one greater than the previous one (e.g., 123, 4567).
 * The constraints on `low` and `high` (up to 10^9) suggest that we cannot iterate through all numbers in the range and check each one.
 * Instead, we can generate all possible sequential digit numbers and then filter them based on the given range.
 *
 * We can generate sequential digit numbers by considering starting digits from 1 to 9 and lengths from 2 up to 9 (since a 10-digit number starting with 1 would exceed 10^9).
 * For each starting digit and length, we can construct the sequential number.
 * For example:
 * - Starting digit 1, length 3: 1 -> 12 -> 123
 * - Starting digit 2, length 4: 2 -> 23 -> 234 -> 2345
 *
 * We can use a queue (or simply an array acting as a queue) for a Breadth-First Search (BFS) like generation.
 * Initialize the queue with single digits 1 through 9.
 * While the queue is not empty:
 *   - Dequeue a number `currentNum`.
 *   - If `currentNum` is within the range [low, high], add it to our result list.
 *   - Determine the last digit of `currentNum`.
 *   - If the last digit is less than 9, construct the next sequential digit number by appending `lastDigit + 1`.
 *   - Enqueue this new number.
 *
 * Finally, sort the collected numbers in ascending order.
 *
 * Alternative generation approach:
 * We can pre-generate all possible sequential numbers (up to 10^9) and store them.
 * The maximum possible sequential number is 123456789.
 * We can iterate through possible lengths of sequential numbers (from 2 to 9).
 * For each length, iterate through possible starting digits (from 1 to 10 - length).
 * Construct the number and add it to a list if it's within the [low, high] range.
 * Sort the final list.
 * This pre-generation approach is simpler to implement and has similar time complexity.
 *
 * Let's use the pre-generation approach for simplicity.
 * The possible sequential numbers are limited and can be generated systematically.
 *
 * Example:
 * Starting digits: 1, 2, 3, 4, 5, 6, 7, 8, 9
 * Possible lengths of sequential numbers:
 * - Length 2: 12, 23, 34, ..., 89
 * - Length 3: 123, 234, ..., 789
 * - ...
 * - Length 9: 123456789
 *
 * We can iterate through all possible starting digits (1-9).
 * For each starting digit, we can build the sequential number by appending the next digit as long as the next digit is <= 9.
 *
 * For example, starting with `1`:
 *   currentNum = 1
 *   nextDigit = 2. currentNum = 12. Check range.
 *   nextDigit = 3. currentNum = 123. Check range.
 *   ...
 *   nextDigit = 9. currentNum = 123456789. Check range.
 *
 * For example, starting with `2`:
 *   currentNum = 2
 *   nextDigit = 3. currentNum = 23. Check range.
 *   nextDigit = 4. currentNum = 234. Check range.
 *   ...
 *   nextDigit = 9. currentNum = 23456789. Check range.
 *
 * We collect all generated numbers that fall within [low, high] and then sort them.
 *
 * @time_complexity
 * Generating all sequential digit numbers is constant time because the maximum number of such integers is limited.
 * The largest sequential digit number is 123456789.
 * The total number of sequential digit integers is relatively small.
 * For a given length `L`, there are `10 - L` possible starting digits.
 * Summing for L from 2 to 9: (10-2) + (10-3) + ... + (10-9) = 8 + 7 + 6 + 5 + 4 + 3 + 2 + 1 = 36.
 * So, there are at most 36 sequential digit numbers to generate.
 *
 * For each generated number, we check if it's within the range [low, high]. This is O(1).
 * Sorting the results: If `R` is the number of sequential digits found within the range, then sorting takes O(R log R). Since `R` is at most 36, this is effectively constant time.
 *
 * Therefore, the overall time complexity is dominated by the generation and sorting of a small, fixed set of numbers, making it O(1) in terms of the input range size.
 *
 * @space_complexity
 * We store the sequential digit numbers that fall within the range. In the worst case, all 36 generated sequential numbers might fall within the range.
 * The space complexity is O(1) because the maximum number of results is bounded by a small constant (36).
 */

/**
 * @param {number} low
 * @param {number} high
 * @return {number[]}
 */
var sequentialDigits = function(low, high) {
    const result = []; // Array to store the sequential digits within the range.
    const digits = "123456789"; // String containing all possible digits to form sequential numbers.

    // Iterate through all possible starting positions in the 'digits' string.
    // 'i' represents the starting digit.
    for (let i = 0; i < digits.length; i++) {
        let currentNum = 0; // Variable to build the current sequential number.

        // Iterate through all possible ending positions in the 'digits' string,
        // starting from the current 'i'.
        // 'j' represents the end of the substring of sequential digits.
        for (let j = i; j < digits.length; j++) {
            // Append the digit at index 'j' to the current number.
            // We convert the character to a number using parseInt() or the unary plus operator (+).
            currentNum = currentNum * 10 + parseInt(digits[j]);

            // Check if the generated currentNum is within the specified range [low, high].
            if (currentNum >= low && currentNum <= high) {
                result.push(currentNum); // If it is, add it to our result list.
            }

            // Optimization: If the current number already exceeds 'high',
            // there's no need to continue building longer sequential numbers
            // starting with this prefix, as they will only be larger.
            // However, this check is implicitly handled by the `currentNum <= high` condition when pushing.
            // A more direct optimization would be:
            // if (currentNum > high) break;
            // but the current loop structure handles it fine by not pushing.
        }
    }

    // The problem requires the output list to be sorted.
    // Since we generate numbers in a somewhat ordered fashion (by starting digit and then length),
    // but not strictly in numerical order (e.g., 123 comes before 23), sorting is necessary.
    result.sort((a, b) => a - b); // Sort the result array in ascending numerical order.

    return result; // Return the sorted list of sequential digits within the range.
};
```